{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-14-804a246b274d>, line 20)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-14-804a246b274d>\"\u001b[1;36m, line \u001b[1;32m20\u001b[0m\n\u001b[1;33m    regex1 =\u001b[0m\n\u001b[1;37m             ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "# Import yo' shiznit\n",
    "\n",
    "import numpy as np\n",
    "import Bio\n",
    "\n",
    "from Bio.Phylo.TreeConstruction import DistanceCalculator\n",
    "from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "from Bio import AlignIO\n",
    "from Bio import Phylo\n",
    "\n",
    "import re\n",
    "import mechanize\n",
    "\n",
    "# First, we have 5 pre-established files containing the FASTA sequence data of 5 sequences from 5 different\n",
    "# phylogenetic tree papers. We use 'mechanize' to open page, and will use form filling to feed the optimizer\n",
    "# the files and extract regexes behind the scenes. Maybe there's a library that can help us with this??\n",
    "\n",
    "br = mechanize.Browser()\n",
    "br.open(\"http://regex.inginf.units.it/\")\n",
    "regex1 = \n",
    "regex2 = \n",
    "regex3 = \n",
    "regex4 = \n",
    "regex5 = \n",
    "................. more to come\n",
    "\n",
    "\n",
    "# This is a sample function designed to read in the name of the user's fasta file\n",
    "# extracted from their regex/resulting sequences\n",
    "\n",
    "def blosumnj(filename):\n",
    "    aln = AlignIO.read(open(filename), 'fasta')\n",
    "    print(aln)\n",
    "\n",
    "    calculator = DistanceCalculator('blosum62')\n",
    "    dm = calculator.get_distance(aln)\n",
    "    print(dm)\n",
    "\n",
    "    from Bio.Phylo.TreeConstruction import DistanceTreeConstructor\n",
    "    constructor = DistanceTreeConstructor(calculator, 'nj')\n",
    "    tree = constructor.build_tree(aln)\n",
    "    print(tree)\n",
    "    \n",
    "# Prompting time! Prompt the users to enter their names, welcome them to the game,\n",
    "# and then to select their desired set of sequences\n",
    "    \n",
    "User1 = (input(\"\\n\\nUser 1: Please enter your name:\\n\\n\"))\n",
    "User2 = (input(\"\\n\\nUser 2: Please enter your name:\\n\\n\"))\n",
    "print(User 1, \"and\", User 2, \"\\nWelcome to the REGEX THROWDOWN!\\n\\n\")\n",
    "\n",
    "print(\"\\nThe objective for this game is to beat your opponent by writing a regular expression whose associated phylogenetic tree most closely matches the phylogenetic tree created by an optimized RegEx created by AI.\")\n",
    "\n",
    "print(\"\\nPlease select which sequence set (based on a phylogenetic tree in a published paper) that you would like to choose for your regex test. Species and name/function of sequence will not be shown.\")\n",
    "\n",
    "print(\"\\nThe blind sequences are as follows:\\n1 - Nazli et al.\\n2 - Audrey et al.\\n3 - Solka et al.\\n 4 - Jafri et al.\\n5 - Jerome et al.\\n\\n\\n\")\n",
    "\n",
    "sequenceselection = int(input(\"\\n\\nPlease enter the number of the sequence set you'd like to choose.\\n\\n\"))\n",
    "\n",
    "if sequenceselection == 1\n",
    "    filename = 'fastafile1.fsa'\n",
    "    else if sequenceselection == 2\n",
    "        filename = 'fastafile2.fsa'\n",
    "    else if sequenceselection == 3\n",
    "        filename = 'fastafile3.fsa'\n",
    "    else if sequenceselection == 4\n",
    "        filename = 'fastafile4.fsa'\n",
    "    else if sequenceselection == 5\n",
    "        filename = 'fastafile5.fsa'\n",
    "    else\n",
    "        print(\"I'm sorry! You need to enter a number between 1 and 5 to indicate your chosen set of sequences. Try again!\")\n",
    "        sequenceselection = int(input(\"\\n\\nPlease enter the number of the sequence set you'd like to choose.\\n\\n\"))    \n",
    "\n",
    "# Prompt the users to enter their best RegEx guesses\n",
    "\n",
    "User1Regex = (input(\"\\n\\nUser 1: Please enter your best guess RegEx.:\\n\\n\"))\n",
    "User2Regex = (input(\"\\n\\nUser 2: Please enter your best guess RegEx.:\\n\\n\"))\n",
    "\n",
    "# The following code block will enter each of the users' RegEx (User1Regex and User2Regex) either as a search term\n",
    "# within ProSite's downloaded file OR on the ProSite/Expasy site iself\n",
    "\n",
    "##### IN THIS CODE BLOCK WE WILL TAKE USER REGEX INPUT, SEARCH PROSITE, AND USE THE RESULTING NUMBERS TO\n",
    "##### GENERATE A FASTA FILE.\n",
    "\n",
    "    # ******** I need to read more on mechanize or other libraries to determine how to proceed to extract ********\n",
    "    # ******** those UniProt ID numbers from the search results                                           ********\n",
    "    \n",
    "    # One promising library: https://github.com/biopython/biopython/blob/master/Bio/ExPASy/ScanProsite.py\n",
    "    \n",
    "    # PROPOSED STRATEGY to convert search results to FSA files:\n",
    "    \n",
    "    # Sample Uniprot search format https://www.uniprot.org/uniprot/?query=SEQ1ID+OR+SEQ2ID+OR+SED3ID+OR+SEQ4ID+OR+SEQ5ID&sort=score\n",
    "    # Sample Uniprot search FASTA result with SEQ1ID,SEQ2ID, etc., limited to first 5 results:\n",
    "    # https://www.uniprot.org/uniprot/?query=SEQ1ID%20OR%20SEQ2ID%20OR%20SEQ3ID%20OR%20%20SEQ4ID%20OR%20SEQ5ID&format=fasta&limit=5&sort=score\n",
    "    \n",
    "    # So what we do, is find a way to extract the top 5 uniprot ids from the prosite motif results, and we store\n",
    "    # the results of those IDs in 5 variables: SEQ1ID, SEQ2ID, etc. (Only the uniprot code instead of these IDs)\n",
    "    \n",
    "    # Then we write a string that is the above Uniprot search result that includes those variable names,\n",
    "    # and extract the website's text, which we then save into a FASTA .fsa file. I might need to try a few\n",
    "    # different ways to successfully create that link. This is all just sketched draft code.\n",
    "    \n",
    "    import bs4\n",
    "    import urllib.request\n",
    "    \n",
    "    link = ('https://www.uniprot.org/uniprot/?query=',SEQ1ID,'%20OR%20',SEQ2ID,'%20OR%20',SEQ3ID,'%20OR%20%20',SEQ4ID,'%20OR%20%20',SEQ5ID,'&format=fasta&limit=5&sort=score')\n",
    "\n",
    "    soup = bs4.BeautifulSoup(webpage)\n",
    "    \n",
    "    print(\"Here is the FASTA file created from the top 5 sequence results generated when your RegEx was run through Scan Prosite's motif search:\")\n",
    "    print(soup.get_text())\n",
    "    \n",
    "    # ******* Insert code block to save the extracted web page text as .fsa file. *******\n",
    "    # ******* I think we can do this with BioPython's SeqIO?                      *******\n",
    "    \n",
    "    # ******* Insert code block to length-equalize the sequences in the .fsa file you just created: *******\n",
    "    # ******* https://www.biostars.org/p/185103/                                                    *******\n",
    "\n",
    "    # This same block of text will extract the uniprot name and/or fasta sequence for the top 5 results.\n",
    "    # We may be able to recruit an existing Python library for that purpose, so we should search a little before hand-coding.\n",
    "\n",
    "    # We also need to find a Python library or tool to generate FASTA files out of subblock A. We will need to\n",
    "    # determine how to equalize the sequences in length, as well, since the BioPython code will not run on sequences of\n",
    "    # unequal length\n",
    "    \n",
    "# Now, we run our blosumnj function on the Regex FASTA file, and the FASTA files we just generated\n",
    "# through the users' regexes.\n",
    "\n",
    "# First create the tree based upon 'filename', which was generated based upon user input of desired sequence\n",
    "\n",
    "blosumnj(filename)\n",
    "\n",
    "# Assign the matrix variable from the function just run to the variable 'exampledm'. Store the tree similarly.\n",
    "\n",
    "exampledm = dm\n",
    "exampletree = tree\n",
    "\n",
    "print(\"Here is the sample distance matrix created with the optimzed RegEx:\\n\\n\", exampledm,\"\\n\\n\")\n",
    "print(\"Here is the tree that results from this sample distance matrix:\\n\\n\", exampletree,\"\\n\\n\")\n",
    "\n",
    "# Now run the blosumnj function on the User 1 FASTA file\n",
    "\n",
    "blosumnj(user1filename)\n",
    "\n",
    "user1dm = dm\n",
    "user1tree = tree\n",
    "\n",
    "print(\"Here is\", User1,\"'s distance matrix:\\n\\n\", user1dm,\"\\n\\n\")\n",
    "print(\"Here is\", User1,\"'s tree:\\n\\n\", user1tree,\"\\n\\n\")\n",
    "\n",
    "# Now run the blosumnj function on the User 2 FASTA file\n",
    "\n",
    "blosumnj(user2filename)\n",
    "\n",
    "user2dm = dm\n",
    "user2tree = tree\n",
    "\n",
    "print(\"Here is\", User2,\"'s distance matrix:\\n\\n\", user2dm,\"\\n\\n\")\n",
    "print(\"Here is\", User2,\"'s tree:\\n\\n\", user2tree,\"\\n\\n\")\n",
    "\n",
    "# Now it is time to score the trees accordingly\n",
    "\n",
    "print(\"It's time to compare and score the players' trees against the optimized RegEx tree!\\n\\nFirst, we will see if all species are the same.\")\n",
    "\n",
    "    # This is where I need some library or other tool or example to figure out how to extract the information from the matrices and trees\n",
    "    # Potential tool: https://github.com/biologyguy/BuddySuite/wiki/PB-Show-unique to show differences between trees, species-wise\n",
    "\n",
    "def score(phylogeny):\n",
    "    \n",
    "    # Here just basically an 'if' loop that matches the species in the RegEx tree with those in the user tree. Basic scoring structure.\n",
    "    \n",
    "    # Another loop that somehow accounts for accuracy of clustering together, if possible.\n",
    "    \n",
    "    # Sum scores, print statement of the score for that user\n",
    "\n",
    "phylogeny = user1tree\n",
    "score(user1tree)\n",
    "user1score = score\n",
    "\n",
    "phylogeny = user2tree\n",
    "score(user2tree)\n",
    "user2score = score\n",
    "\n",
    "if user1score>user2score\n",
    "    winner = User1\n",
    "    print(User1\"'s regex score is,\" user1score,\"! \"User2\"'s RegEx score is,\" user2score,\"! \", User1, \"you CRUSHED THE COMPETITION! LET THE BRAGGING RIGHTS COMMENCE!\\n\\n\")\n",
    "        else if user2score>user1score\n",
    "            winner = User2\n",
    "            print(User2\"'s regex score is,\" user2score,\"! \"User1\"'s RegEx score is,\" user1score,\"! \", User2, \"you CRUSHED THE COMPETITION! LET THE BRAGGING RIGHTS COMMENCE!\\n\\n\")\n",
    "\n",
    "    # OR: Another VERY promising comparison tool that seems to do it all: http://etetoolkit.org/documentation/ete-compare/\n",
    "    # http://etetoolkit.org/docs/latest/tutorial/tutorial_phylogeny.html?highlight=fasta#linking-phylogenetic-trees-with-multiple-sequence-alignments\n",
    "                       \n",
    "print(winner,\", YOU GET TO VICTORY DANCE WITH CARLTON!\")\n",
    "br.open(\"https://youtu.be/Lxqa2Haf8lo?t=19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
